File: proyecto1/src/main/Modelo/Mensaje/Mensaje.go
package mensaje

import "strings"

//Mensaje será el programa del lado del servidor que nos ayude a
//manejar los mensajes. Guardará la información necesaria del
//mensaje y nos permitirá acceder a esta

//Mensaje con los campos posibles que puede tener el mensaje JSON.
type Mensaje struct{
	Tipo string `json:"type,omitempty"`
	Username string `json:"username,omitempty"`
	Operation string `json:"operation,omitempty"`
	Result string `json:"result,omitempty"`
	Extra string `json:"extra,omitempty"`
	Status string `json:"status,omitempty"`
	Users map[string]string `json:"users,omitempty"`
	Usernames []string `json:"usernames,omitempty"`
	Text string `json:"text,omitempty"`
	Roomname string `json:"roomname,omitempty"`
}

func CrearMensajeIdentify(username string) *Mensaje {
	return &Mensaje{
		Tipo : "IDENTIFY",
		Username : username,
	}
}

func CrearMensajeResponse(operation, result, extra string) *Mensaje {
	return &Mensaje{
		Tipo : "RESPONSE",
		Operation : operation,
		Result : result,
		Extra : extra,
	}
}

func CrearMensajeNewUser(username string) *Mensaje {
	return &Mensaje{
		Tipo : "NEW_USER",
		Username : username,
	}
}

func CrearMensajeStatus(status string) *Mensaje {
	return &Mensaje{
		Tipo : "STATUS",
		Status : status,
	}
}

func CrearMensajeNewStatus(username, status string) *Mensaje {
	return &Mensaje{
		Tipo : "NEW_STATUS",
		Username : username,
		Status : status,
	}
}

func CrearMensajeUsers() *Mensaje {
	return &Mensaje{
		Tipo : "USERS",
	}
}

func CrearMensajeUserList(users map[string]string) *Mensaje {
	return &Mensaje{
		Tipo : "USER_LIST",
		Users : users,
	}
}

func CrearMensajeText(username, text string) *Mensaje {
	return &Mensaje{
		Tipo : "TEXT",
		Username : username,
		Text : text,
	}
}

func CrearMensajeTextFrom(username, text string) *Mensaje {
	return &Mensaje{
		Tipo : "TEXT_FROM",
		Username : username,
		Text : text,
	}
}

func CrearMensajePublicText(text string) *Mensaje {
	return &Mensaje{
		Tipo : "PUBLIC_TEXT",
		Text : text,
	}
}

func CrearMensajePublicTextFrom(username, text string) *Mensaje {
	return &Mensaje{
		Tipo : "PUBLIC_TEXT_FROM",
		Username : username,
		Text : text,
	}
}

func CrearMensajeNewRoom(roomname string) *Mensaje {
	return &Mensaje{
		Tipo : "NEW_ROOM",
		Roomname : roomname,
	}
}

func CrearMensajeInvite(roomname string, usernames []string) *Mensaje {
	return &Mensaje{
		Tipo : "INVITE",
		Roomname : roomname,
		Usernames : usernames,
	}
}

func CrearMensajeInvitation(username, roomname string) *Mensaje {
	return &Mensaje{
		Tipo : "INVITATION",
		Username : username,
		Roomname : roomname,
	}
}

func CrearMensajeJoinRoom(roomname string) *Mensaje {
	return &Mensaje{
		Tipo : "JOIN_ROOM",
		Roomname : roomname,
	}
}

func CrearMensajeJoinedRoom(roomname, username string) *Mensaje {
	return &Mensaje{
		Tipo : "JOINED_ROOM",
		Roomname : roomname,
		Username : username,
	}
}

func CrearMensajeRoomUsers(roomname string) *Mensaje {
	return &Mensaje{
		Tipo : "ROOM_USERS",
		Roomname : roomname,
	}
}

func CrearMensajeRoomText(roomname, text string) *Mensaje {
	return &Mensaje{
		Tipo : "ROOM_TEXT",
		Roomname : roomname,
		Text : text,
	}
}

func CrearMensajeRoomTextFrom(roomname, username, text string) *Mensaje {
	return &Mensaje{
		Tipo : "ROOM_TEXT_FROM",
		Roomname : roomname,
		Username : username,
		Text : text,
	}
}

func CrearMensajeLeaveRoom(roomname string) *Mensaje {
	return &Mensaje{
		Tipo : "LEAVE_ROOM",
		Roomname : roomname,
	}
}

func CrearMensajeLeftRoom(roomname, username string) *Mensaje {
	return &Mensaje{
		Tipo : "LEFT_ROOM",
		Roomname : roomname,
		Username : username,
	}
}

func CrearMensajeDisconnect() *Mensaje {
	return &Mensaje{
		Tipo : "DISCONNECT",
	}
}

func CrearMensajeDisconnected(username string) *Mensaje {
	return &Mensaje{
		Tipo : "DISCONNECTED",
		Username : username,
	}
}

func (msg *Mensaje)EsValido() bool {
	return strings.TrimSpace(msg.Tipo) != ""
}

func (msg *Mensaje)GetTipo() string {
	return msg.Tipo
}

func (msg *Mensaje)GetUsername() string {
	return msg.Username
}

func (msg *Mensaje)GetOperation() string {
	return msg.Operation
}

func (msg *Mensaje)GetResult() string {
	return msg.Result
}

func (msg *Mensaje)GetExtra() string {
	return msg.Extra
}

func (msg *Mensaje)GetStatus() string {
	return msg.Status
}

func (msg *Mensaje)GetUsers() map[string]string {
	return msg.Users
}

func (msg *Mensaje)GetUsernames() []string {
	return msg.Usernames
}

func (msg *Mensaje)GetText() string {
	return msg.Text
}

func (msg *Mensaje)GetRoomname() string {
	return msg.Roomname
}

File: proyecto1/src/main/Modelo/Mensaje/Mensaje.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Collections.Generic;

namespace mensaje{

    public class Mensaje{
	[JsonPropertyName("type")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? tipo {get;set;}
	
	[JsonPropertyName("username")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? username {get;set;}
	
	[JsonPropertyName("operation")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? operation {get;set;}
	
	[JsonPropertyName("result")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? result {get;set;}
	
	[JsonPropertyName("extra")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? extra {get;set;}
	
	[JsonPropertyName("status")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? status {get;set;}
	
	[JsonPropertyName("users")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public Dictionary<string, string>? users {get;set;}
	
	[JsonPropertyName("usernames")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public List<string>? usernames {get;set;}
	
	[JsonPropertyName("text")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? text {get;set;}
	
	[JsonPropertyName("roomname")]
	[JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
	public string? roomname {get;set;}
    
	public static Mensaje CrearMensajeIdentify(string username){
	    return new Mensaje{tipo = "IDENTIFY", username = username};
	}

	public static Mensaje CrearMensajeResponse(string operation, string result, string extra){
	    return new Mensaje{tipo = "RESPONSE", operation = operation, result = result, extra = extra};
	}

	public static Mensaje CrearMensajeNewUser(string username){
	    return new Mensaje{tipo = "NEW_USER", username = username};
	}

	public static Mensaje CrearMensajeStatus(string status){
	    return new Mensaje{tipo = "STATUS", status = status};
	}

	public static Mensaje CrearMensajeNewStatus(string username, string status){
	    return new Mensaje{tipo = "NEW_STATUS", username = username, status = status};
	}

	public static Mensaje CrearMensajeUsers(){
	    return new Mensaje{tipo = "USERS"};
	}

	public static Mensaje CrearMensajeUserList(Dictionary<string, string> users){
	    return new Mensaje{tipo = "USER_LIST", users = users};
	}

	public static Mensaje CrearMensajeText(string username, string text){
	    return new Mensaje{tipo = "TEXT", username = username, text = text};
	}

	public static Mensaje CrearMensajeTextFrom(string username, string text){
	    return new Mensaje{tipo = "TEXT_FROM", username = username, text = text};
	}

	public static Mensaje CrearMensajePublicText(string text){
	    return new Mensaje {tipo = "PUBLIC_TEXT", text = text };
	}

	public static Mensaje CrearMensajePublicTextFrom(string username, string text){
	    return new Mensaje {tipo = "PUBLIC_TEXT_FROM", username = username, text = text };
	}

	public static Mensaje CrearMensajeNewRoom(string roomname){
	    return new Mensaje {tipo = "NEW_ROOM", roomname = roomname };
	}

	public static Mensaje CrearMensajeInvite(string roomname, List<string> usernames){
	    return new Mensaje {tipo = "INVITE", roomname = roomname, usernames = usernames };
	}

	public static Mensaje CrearMensajeInvitation(string username, string roomname){
	    return new Mensaje {tipo = "INVITATION", username = username, roomname = roomname };
	}

	public static Mensaje CrearMensajeJoinRoom(string roomname){
	    return new Mensaje {tipo = "JOIN_ROOM", roomname = roomname };
	}

	public static Mensaje CrearMensajeJoinedRoom(string roomname, string username){
	    return new Mensaje {tipo = "JOINED_ROOM", roomname = roomname, username = username };
	}

	public static Mensaje CrearMensajeRoomUsers(string roomname){
	    return new Mensaje {tipo = "ROOM_USERS", roomname = roomname };
	}

	public static Mensaje CrearMensajeRoomText(string roomname, string text){
	    return new Mensaje {tipo = "ROOM_TEXT", roomname = roomname, text = text };
	}

	public static Mensaje CrearMensajeRoomTextFrom(string roomname, string username, string text){
	    return new Mensaje {tipo = "ROOM_TEXT_FROM", roomname = roomname, username = username, text = text };
	}

	public static Mensaje CrearMensajeLeaveRoom(string roomname){
	    return new Mensaje {tipo = "LEAVE_ROOM", roomname = roomname };
	}

	public static Mensaje CrearMensajeLeftRoom(string roomname, string username){
	    return new Mensaje {tipo = "LEFT_ROOM", roomname = roomname, username = username };
	}

	public static Mensaje CrearMensajeDisconnect(){
	    return new Mensaje {tipo = "DISCONNECT" };
	}

	public static Mensaje CrearMensajeDisconnected(string username){
	    return new Mensaje {tipo = "DISCONNECTED", username = username };
	}

	public bool esValido(){
	    return !string.IsNullOrWhiteSpace(tipo);
	}
    }
}
    
